classes/class-p4-taxonomy-image.php

```diff
@@ -0,0 +1,148 @@
+		/**
+		 * Taxonomy_Image constructor.
+		 */
+		public function __construct() {
+			add_action( 'post_tag_add_form_fields', array( $this, 'tag_form_fields' ) );
+			add_action( 'post_tag_edit_form_fields', array( $this, 'tag_form_fields' ) );
+			add_action( 'create_post_tag', array( $this, 'save_tag_meta' ) );
+			add_action( 'edit_post_tag', array( $this, 'save_tag_meta' ) );
+			add_action( 'admin_enqueue_scripts', array( $this, 'enqueue_admin_assets' ) );
+
+			add_filter( 'manage_edit-post_tag_columns', array( $this, 'edit_post_tag_columns' ) );
+			add_filter( 'manage_post_tag_custom_column', array( $this, 'manage_post_tag_custom_column' ), 10, 3 );
+			add_filter( 'manage_edit-post_tag_sortable_columns', array( $this, 'manage_post_tag_custom_sortable_column' ), 10, 3 );
+		}
+
+		/**
+		 * Add/Edit Tag page custom field template.
+		 *
+		 * @param WP_Term $wp_tag The object passed to the callback when on Edit Tag page.
+		 */
+		public function tag_form_fields( $wp_tag ) {
+			if ( isset( $wp_tag ) && $wp_tag instanceof WP_Term ) { ?>
+				<tr class="form-field term-image-wrap">
+					<th>
+						<label for="tag_attachment_id"><?php esc_html_e( 'Image', 'planet4-master-theme' ); ?></label>
+					</th>
+					<td>
+						<input type="hidden" name="tag_attachment_id" id="tag_attachment_id" class="tag_attachment_id" value="" />
+						<button class="button insert-media add_media" name="insert_image_tag_button" id="insert_image_tag_button" type="button">
+							<?php esc_html_e( 'Select/Upload Image', 'planet4-master-theme' ); ?>
+						</button>
+						<p class="description"><?php esc_html_e( 'Associate this tag with an image.', 'planet4-master-theme' ); ?></p>
+						<p class="tag_attachment">
+							<?php echo wp_get_attachment_image( get_term_meta( $wp_tag->term_id, 'tag_attachment_id', true ) ); ?>
+						</p>
+					</td>
+				</tr>
+			<?php } else { ?>
+				<div class="form-field term-image-wrap">
+					<label for="tag_attachment_id"><?php esc_html_e( 'Image', 'planet4-master-theme' ); ?></label>
+					<input type="hidden" name="tag_attachment_id" id="tag_attachment_id" class="tag_attachment_id" value="" />
+					<button class="button insert-media add_media" name="insert_image_tag_button" id="insert_image_tag_button" type="button">
+						<?php esc_html_e( 'Select/Upload Image', 'planet4-master-theme' ); ?>
+					</button>
+					<p class="description"><?php esc_html_e( 'Associate this tag with an image.', 'planet4-master-theme' ); ?></p>
+					<p class="tag_attachment">
+						<img class="attachment-thumbnail size-thumbnail" src=""/>
+					</p>
+				</div>
+				<?php
+			}
+		}
+
+		/**
+		 * Save user custom input field.
+		 *
+		 * @param int $term_id The ID of the WP_Term object that is added or edited.
+		 */
+		public function save_tag_meta( $term_id ) {
+			$attachment_id = filter_input( INPUT_POST, 'tag_attachment_id', FILTER_VALIDATE_INT );
+			if ( $attachment_id ) {
+				$attachment_id = $this->validate( $attachment_id );
```

Comment:
Not sure we need `$this->validate` here. The `filter_input` above validates if it's int and converts to int on success ref: http://php.net/manual/en/filter.filters.validate.php

Also if ( $attachment_id ) above will fail on int=0.

So, I'd remove this line and corresponding `validate()` function as it's not used anywhere else. Also remove the `if` statement below.

Reply:
The validate function does not validate if it's int. It validates if its positive int.
Failing on 0 was ok, that is why I did not care about it.
But now I changed the function to return only bool.

Reply:
@angtheod Hm, I'm still a bit confused here. The `validate()` will return `true` if it's `int and it's >0`. 

So `filter_input()` will convert value to int, so in `if` statement we can just check if it's `>0`:
`if ( $attachment_id > 0 ) {`

I don't see the value of separating this logic into separate method if it's used just ones, and it's clear to see the check closer to the code where it's performed. If that makes sense?

Reply:
Actually here I wonder if a check to see if term actually exist would be a good one, as anything can be passed in `$_POST`

Reply:
You mean the 'tag_attachment_id' key in $_POST? If this is not set then filter_input will return NULL or if the filter fails for whatever reason it will return false.
In both cases the if statement inside validate() results to True (since NULL == 0 and false == 0 are both True), so validation will not pass.

As for checking this in a separate function, I guess it is just a matter of preference... I know it is a trivial function, but since I put this all inside a class I prefer to have a separate method for validation logic just to know at first glance that this class handles also validation.